syntax = "proto3";

// Buy at any price given amount on assets
message BuyOrder {
  // who order
  string clientId = 1;
  // What order
  string ticker = 2;
  // order option
  oneof option {
    BuyMax buyMax = 3;
    BuyOne buyOne = 4;
  }
}

// buy as much assets as you can based on available budget
message BuyMax {
}

// buy only one asset
message BuyOne {
}

// Sell at any price given amount on assets
message SellOrder {
  // who order
  string clientId = 1;
  // What order
  string ticker = 2;
  // how many items
  double amount = 3;
}

message OrderBook {
}

message Finished {
}

message State {
  OrderBook orderBook = 1;
  Finished finished = 2;
  double budget = 3;
}


service Agent {
    rpc buy(BuyOrder) returns (State) {}
    rpc sell(SellOrder) returns (State) {}
}
